### PR TITLE
[chrome] Make non-promisified signature require a callback

### PR DESCRIPTION
**Pull request by fregante**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

---

This is an overall definition cleanup. Here's the history of the types until now:

1. chrome.* APIs had an optional callback and returned `void`: `do(callback?: Function): void`
2. since MV3, omitting the callback causes the API to return `Promise<Foo>`
3. types for this promisified signature were added, but the previous ones were not updated

This means that many of these signatures now return `Foo | Promise<Foo>` if the callback is omitted, which isn't quite right and confuses some lint rules. In some cases, if the `callback` overload comes before the Promise, the latter signature is ignored and the return type is just `Foo`.

Avoiding `callback?` everywhere avoids these order issues and reflect the API more closely

Example from the documentation: https://developer.chrome.com/docs/extensions/reference/action/#method-getTitle

> In Manifest V3 and later, return a Promise by omitting the callback argument.

### CODE REVIEW

File: types/chrome/index.d.ts [hunk 1]
```diff
@@ -167,7 +167,7 @@ declare namespace chrome.action {
      * @param tabId The id of the tab for which you want to modify the action.
      * @param callback
      */
-    export function disable(tabId?: number, callback?: () => void): void;
+    export function disable(tabId?: number, callback: () => void): void;
```
Comment (fregante) [inline types/chrome/index.d.ts:170]: Reviewer note: These changes were made only on methods that already had a Promise-returning definition.

**Comment by matcarlson** [comment]
Can you give an example of the warning/error that you've seen, and the `chrome.*` call that causes it?

**Review by fregante — COMMENTED** [review]
> Can you give an example of the warning/error that you've seen, and the `chrome.*` call that causes it?

I don't remember which API was in particular, but overloads are matched in order, so if you have this type:

```ts
function get(callback?: (value: string) => void): void;
function get(): Promise<string>;
```

The value of `get()` will be `void | Promise<string>`. I wrote this in my first post as well.

https://www.typescriptlang.org/play?target=9#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwHMQMAKMKCCAIyjAGsB+ALnhIDcLkQWBnDGLKgIBKeAF4AfPDY4swYSxlyA3AChQkWAhTpseQsRIL4ABRg4Atlh4gAPHwFCJa1WDx9pnbvAeCC4+CgAdygsDANSYRd1cGg4JDRMXHwiDAAxLAAPEGAyCmpaOhZ2L15+P1FJaVl5RRq1DTjtRL0U4gzs3OMzS2s7XycXN1QPDgguDpyyx38xQJCwiMmu6NU19Y31kEyABxwYcIBvMa4AGk9xkGWAXyA

Screenshot 3


To be fair, most of these are in the right order, but there's no reason to keep `callback?` in the types. It just causes confusion and can cause this issue when people promisify the rest of the API by looking at existing types.

File: types/chrome/index.d.ts [hunk 2]
```diff
@@ -1178,19 +1184,19 @@ declare namespace chrome.browsingData {
...
      * function() {...};
      */
-    export function removePasswords(options: RemovalOptions, callback?: () => void): void;
+    export function removePasswords(options: RemovalOptions, callback: () => void): void;
     /**
      * Clears the browser's cookies and server-bound certificates modified within a particular timeframe.
      * @return The `removeCookies` method provides its result via callback or returned as a `Promise` (MV3 only). It has no parameters.
      */
-    export function removeCookies(options: RemovalOptions, callback?: () => void): Promise<void>;
+    export function removeCookies(options: RemovalOptions): Promise<void>;
```
[the remark this answers is not in the document]
Comment (matcarlson) [reply types/chrome/index.d.ts:1192]: Sorry, I missed this in the first comment. 

IMO this looks reasonable so far, but it'll take me a bit to get through it. I'll work on it over the weekend.

**Review by matcarlson — APPROVED** [review]
Sorry for the delay, I think I worked on your other PR and forgot about this one. Thanks for fixing these!

**Review by matcarlson — APPROVED** [review]

**Comment by fregante** [comment]
Ready to merge

**Comment by fregante** [comment]
Thanks @matcarlson! I hope to contribute more to types/chrome when I have time

**Comment by bershanskiy** [comment]
I believe this commit broke promise variant of `browserAction.setIcon()`. Specifically, the following code is no longer valid:
```js
chrome.browserAction.setIcon({ path: '/some/path.png' });
```

I believe the following declaration should be added somewhere around line 921:
```js
export function setIcon(details: TabIconDetails): Promise<void>;
```

I could not find a bug report or a PR yet, I will try to create a PR at some point if I don't find one.

**Comment by fregante** [comment]
I think it was a bad merge with https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64235 😰

**Comment by bershanskiy** [comment]
I looked into this a bit more and realized that:

 - only Firefox exposes both calling conventions on both `browser` and `chrome` objects (which are the same object).
image
 - Chrome allows calling these methods without callbacks (callbacks are optional).

So then the optional callback and `void` return would work for Chrome, and no callback and `Promise<void>` return would work for Firefox.

**Comment by fregante** [comment]
I don’t think this package cares about Firefox, otherwise it would have returned Promise a long time ago. Firefox has its own types

**Comment by bershanskiy** [comment]
@fregante I finally got around to fixing the `chrome.browserAction` namespace, the fix is in #64604.

**Pull request merged**